[MLIR][LLVMIR][DLTI] Add `LLVM::TargetAttrInterface` and `#llvm.target` attr

rolfmorel opened the pull request:
Adds the `#llvm.target<triple = $TRIPLE, chip = $CHIP, features = $FEATURES>` attribute and along with a `-llvm-target-to-data-layout` pass to derive a MLIR data layout from the LLVM data layout string (using the existing `DataLayoutImporter`). The attribute implements the relevant DLTI-interfaces, to expose the `triple`, `chip` (AKA `cpu`) and `features` on `#llvm.target` and the full `DataLayoutSpecInterface`. The pass combines the generated `#dlti.dl_spec` with an existing `dl_spec` in case one is already present, e.g. a `dl_spec` which is there to specify size of the `index` type.

Adds a `TargetAttrInterface` which can be implemented by all attributes representing LLVM targets.

Similar to the Draft PR https://github.com/llvm/llvm-project/pull/78073.

RFC on which this PR is based: https://discourse.llvm.org/t/mandatory-data-layout-in-the-llvm-dialect/85875

Commit:
[MLIR][LLVMIR][DLTI] Add LLVMTargetAttrInterface, #llvm.target and #llvm.data_layout

Commit:
Try to fix Flang...

Commit:
Fix formatting

rengolin commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -13,13 +13,118 @@ include "mlir/Dialect/LLVMIR/LLVMDialect.td"
+// LLVM_TargetAttr
+//===----------------------------------------------------------------------===//
+
+def LLVM_TargetAttr : LLVM_Attr<"Target", "target", [LLVM_TargetAttrInterface]> {
```
One problem we faced on our side earlier was that it's not that easy to know you're constructing an invalid `TargetMachine` ([link](https://discourse.llvm.org/t/generating-riscv-code-from-mlir/85863)). But validating all possible combinations is just not viable. For downstream targets, you'd have to overload some verification to match their alternatives.

Other tools don't verify much either (other than completely crash if the strings are invalid), so I guess we can start with that. The discussion about which combinations are valid ([link](https://discourse.llvm.org/t/move-remove-vcix-dialect/86920/46)) are generally higher in the pipeline (front-end). At the IR level, we can only treat them as "gospel" and crash if we really can't build a `TargetMachine`.

Commit:
Minor fixes

Commit:
Add first two tests

Commit:
Update docs

Commit:
Add cmake comments and a test

rolfmorel commented:
Couple of pain points in need of suggestions/feedback:

1. As `LLVMTargetFeaturesAttr` exists we have the following:
   1. There's a circular dependency between `LLVMInterfaces` and `LLVMAttrs` in that the new `TargetAttrInterface` requires `LLVMTargetFeaturesAttr` (for the `features` of a target) and there are attrs in `LLVMAttrs` that require interfaces from `LLVMInterfaces`. I break the dependency by moving `LLVMTargetFeaturesAttr` to its own file and having a dedicated CMake-rule to generate `LLVMOpsAttrDefs.h.inc` by way of `LLVMAttrAndEnumDefs.td`.
        1. Is there a better way of dealing with this circular dependency? 
   3. Currently, `#nvvm.target` and `#rocdl.target` take `features` as a `StringAttr`. Seems to me they should take `LLVMTargetFeaturesAttr` which would make them eligible for/compatible with implementing `TargetAttrInterface`. I have not spent time on transitioning them yet.
4.  For some reason, the parser generated for ``let assemblyFormat = [{`<` `triple` `=` $triple (`,` `chip` `=` $chip^)? (`,` qualified($features)^)? `>`}];`` does not actually make the `chip` clause optional. That is, it greedily assumes that as it saw `,` from that clause and it is now committed to it. For example, `#llvm.target<triple = "x86_64-unknown-linux", #llvm.target_features<["+mmx", "+sse"]>>` yields an `expected 'chip'` parser error.
5. How do I properly hang a `unique_ptr<llvm::TargetMachine>` of off `TargetAttr`? The current bare pointer works but probably is not safe, e.g. in case the context gets deleted. See discussion with @rengolin above.
6. Had to move the `DataLayoutImporter` from `lib/Target/LLVMIR` to `Dialect/LLVMIR/IR`, in support of `#llvm.data_layout` being a wrapper around the corresponding `#dlti.dl_spec`, which causes a linking dependency on the DLTI-dialect.
7. What is the strategy around verification? See discussion with @rengolin above.
8. Test cases will need to be made aware of which backends the compiled `llvm-project` has available and be skipped if not applicable.

fabianmcg reviewed (commented):
Nice! Quick comment, see my comment on `LLVMAttrs.h`, that affects many things in this PR and should be fixed. I'll give it another look later in the week.

rolfmorel commented:
@fabianmcg thanks for pointing that out! That should resolve bullet 4 up above I believe.

Any other feedback on this PR would be much appreciated. In particular any help with addressing the remaining bullets:

> 1. As the `LLVMTargetFeaturesAttr` attribute already exists we have the following:
>    
>    1. There's a circular dependency between `LLVMInterfaces` and `LLVMAttrs` in that the new `TargetAttrInterface` requires `LLVMTargetFeaturesAttr` (for the `features` of a target) and there are attrs in `LLVMAttrs` that require interfaces from `LLVMInterfaces`. I break the dependency by moving `LLVMTargetFeaturesAttr` to its own file and having a dedicated CMake-rule to generate `LLVMOpsAttrDefs.h.inc` by way of `LLVMAttrAndEnumDefs.td`.
>       
>       1. Is there a better way of dealing with this circular dependency?
>    2. Currently, `#nvvm.target` and `#rocdl.target` take `features` as a `StringAttr`. Seems to me they should take `LLVMTargetFeaturesAttr` which would make them eligible for/compatible with implementing `TargetAttrInterface`. I have not spent time on transitioning them yet.
> 2. For some reason, the parser generated for `` let assemblyFormat = [{`<` `triple` `=` $triple (`,` `chip` `=` $chip^)? (`,` qualified($features)^)? `>`}]; `` does not actually make the `chip` clause optional. That is, it greedily assumes that as it saw `,` from the `chip` clause it is now committed to it. For example, `#llvm.target<triple = "x86_64-unknown-linux", #llvm.target_features<["+mmx", "+sse"]>>` yields an `expected 'chip'` parser error.
>    
>    1. What's the best solution here? Just give in and write a custom parser?
> 3. How do I properly hang a `unique_ptr<llvm::TargetMachine>` of off `TargetAttr`? The current bare pointer works but probably is not safe, e.g. in case the context gets deleted. See discussion with @rengolin above.
> 
> 4. ~Had to move the `DataLayoutImporter` from `lib/Target/LLVMIR` to `Dialect/LLVMIR/IR`, in support of `#llvm.data_layout` being a wrapper around the corresponding `#dlti.dl_spec`, which causes a linking dependency on the DLTI-dialect.~
> 
> 5. What is the strategy around verification? See discussion with @rengolin above.
> 6. Test cases should be made aware of which backends the compiled `llvm-project` has available and be skipped if not applicable. CI is happy for now, which is actually wrong.

rengolin commented:
> Also, I'm not sure whether making `chip` optional is a good idea.

Older Arm targets didn't need a chip because the micro-architecture was encoded in the target triple. There can be the case where passing a chip name is not just confusing but wrong (if the target doesn't recognize the chip but can build a target with the correct set of features). For x86 and GPUs the chip names carry a lot of information, that is not encoded in the triple, so become necessary.

Arguably we don't usually care about armv7 or lower, but I'll let @banach-space comment on that.

fabianmcg commented:
>  The cost model will be making a lot of enquires about the target over the life of the compiler passes, so this cache is necessary to avoid constructing it for every invocation.

Do you already have the design of how that looks? I'm not saying it's not necessary, I'm arguing that until such use is ready to land we shouldn't put it in, as we would be adding unused features, and it's easy to add later when the use is ready.

rengolin commented:
> In that particular case, I'd prefer an empty chip or a string indicating this is a special case. Because as you say, there are a lot of targets where the chip does carry a lot info, therefore I think we should handle the common case of required chip.

Empty string would be fine, yes. If Arm is fine with making it compulsory (and not empty), we should. Validation would be so much easier. 

If not, then we may need some custom validation per target to make sure that for the ones where the info is in the chip, it must not be empty.

rengolin commented:
> Do you already have the design of how that looks? I'm not saying it's not necessary, I'm arguing that until such use is ready to land we shouldn't put it in, as we would be adding unused features, and it's easy to add later when the use is ready.

I'd be ok with that, too.

Commit:
Make chip required, no caching of TargetMachine, and take features as a string

Commit:
Resolve issues based on review

Issue 1. - features is now a StringAttr
Issue 2. - use struct for attribute parsing
Issue 3. - forgo trying to cache llvm::TargetMachine
Issue 4. - keep code that needs to link to llvm libs in mlir/Target

Commit:
ModuleImport fixes

Commit:
Fixes and cleanup and target dependence for test cases

Commit:
Remove commented code

fabianmcg reviewed (commented):
I gave it a quick look, will inspect more later.

rolfmorel commented on mlir/lib/Target/LLVM/Target.cpp:
```diff
@@ -0,0 +1,89 @@
+  FailureOr<std::unique_ptr<llvm::TargetMachine>>
+  getTargetMachine(Attribute attribute) const {
+    auto attr = llvm::cast<LLVM::TargetAttr>(attribute);
+
+    llvm::InitializeAllTargets();
+    llvm::InitializeAllTargetMCs();
+
```
My understanding is that this is like tools like `mlir-opt` registering all the dialects and their extensions:
https://github.com/llvm/llvm-project/blob/ef98e248c7740fb882b256dd325d22a057de1951/llvm/include/llvm/Support/TargetSelect.h#L58-L89

If you have a concrete suggestion for how to "initialize" just the relevant backends, I am all ears.

fabianmcg replied:
Ok, I'd argue that a better approach would be having a target attribute for each LLVM target, that way only used targets get initialized.

fabianmcg replied:
We do have one: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVM/NVVM/Target.cpp#L124-L129

If each target has it's own attribute then the problem is solved.

Another option, is to add a target name parameter to the attribute, and init only that target.

fabianmcg replied:
To be honest I'd like to see a dialect per target, and also an attribute per target, that would provide stronger target semantics. But, that's not something we need to solve in this PR.

In this case, since the pass is going to be moved to `target` and a default behavior for testing might be good, how about if we add the call to the pass but guarded with a macro and a CMake option? that way people can opt out and use other mechanisms to initialize the targets they care about. As a temporary measure until we get better target support for all targets.

rolfmorel commented on mlir/include/mlir/Dialect/LLVMIR/DataLayoutImporter.h:
```diff
@@ -11,8 +11,8 @@
 
-#ifndef MLIR_LIB_TARGET_LLVMIR_DATALAYOUTIMPORTER_H_
-#define MLIR_LIB_TARGET_LLVMIR_DATALAYOUTIMPORTER_H_
+#ifndef MLIR_DIALECT_LLVMIR_DATALAYOUTIMPORTER_H
```
So we can have a `#llvm.data_layout<DATA_LAYOUT_STR>` attribute. The DATA_LAYOUT_STR is parsed by this importer and yields an unwieldly `#dlti.dl_spec<...........................................................................................................................................................................................................................................................................................................................>`. What the attribute does is to just show the compact encoding, i.e. the string.

fabianmcg replied:
In general it doesn't matter if the textual version of the IR looks ugly, it's IR. It's preferable to have only one way to specify things, and in this case I'd argue that the verbose approach is better.

Re the second comment. You can make the Attr interface return the DL attribute, that way the interface impl and the importer still live in target.

joker-eph replied:
Side note: why is the DLTI using dense attributes for storing integers?
The example above would be much shorter with simple integers I think.

rolfmorel commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrs.h:
```diff
@@ -14,12 +14,15 @@
 
 #include "mlir/Dialect/LLVMIR/LLVMOpsEnums.h.inc"
 
+#include "llvm/IR/DataLayout.h"
+#include "llvm/Target/TargetMachine.h"
```
As the `mlir::LLVM::TargetAttrInterface` has methods that return the DataLayout and TargetMachine.

Look at `mlir/include/mlir/Dialect/LLVMIR/LLVMDialect.h` to see that having dependency on LLVM's headers is not a big deal. For clarity's sake, due moving the parts that require linking to `mlir/Target` the `LLVM` dialect does not gain any linking dependencies.

fabianmcg replied:
Actually those dependencies there are spurious and legacy, see https://github.com/llvm/llvm-project/pull/150692 for removal

Ideally the LLVM dialect shouldn't depend on `llvm/IR`.

rolfmorel replied:
Will have another go at getting these bits to live in `mlir/lib/Target`.

rolfmorel commented on mlir/lib/Dialect/LLVMIR/Transforms/DataLayoutFromTarget.cpp:
```diff
@@ -0,0 +1,61 @@
+
+    bool passFailed = false;
+
+    mod->walk([&](ModuleOp mod) {
```
Fair. Should it be `Operation` as that's the only common ancestor between `builtin::ModuleOp` and `gpu::ModuleOp`?

joker-eph replied:
I would use an Operation pass, and not perform any walk but just lookup the current operation and get a LLVMTargetAttr directly from there.

fabianmcg commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -1304,6 +1305,128 @@ def LLVM_TargetFeaturesAttr : LLVM_Attr<"TargetFeatures", "target_features">
+// LLVM_DataLayoutAttr
+//===----------------------------------------------------------------------===//
+
+def LLVM_DataLayoutAttr
```
Yeah, as above, I would argue verbose is better, it doesn't have to look pretty.

joker-eph commented on mlir/lib/Dialect/LLVMIR/Transforms/DataLayoutFromTarget.cpp:
```diff
@@ -0,0 +1,61 @@
+
+    mod->walk([&](ModuleOp mod) {
+      auto targetAttr =
+          mod->getAttrOfType<LLVM::TargetAttrInterface>("llvm.target");
```
We shouldn't have a magic strings, can you define "llvm.target" as a `let discardableAttrs = (ins ...` on the dialect instead?

joker-eph commented on mlir/lib/Target/LLVM/Target.cpp:
```diff
@@ -0,0 +1,89 @@
+void LLVM::registerLLVMTargetInterfaceExternalModels(
+    DialectRegistry &registry) {
+  registry.addExtension(+[](MLIRContext *ctx, LLVM::LLVMDialect *dialect) {
+    LLVM::TargetAttr::attachInterface<LLVMTargetAttrImpl>(*ctx);
```
I would 
1) remove the interface
2) Make the pass operate directly on LLVMTargetAttr.
3) Move the implementation of the interface as private functions  in the pass (getting DataLayout and TargetMachine from a LLVMTargetAttr).

That will make the pass itself dependent on the LLVM backend, but nothing else.

fabianmcg replied:
> That means the interface here will already be linked and loaded whenever the LLVMTargetAttr will be used. Which forces to link the LLVM backend, which does not seem desirable.

If the interface is never used (ie. appears in a `cast`) then the promise never asserts and one doesn't have to register the interface, so in theory it only introduces an optional dependency, and a har dep on users of the interface, or did the mechanism change?

Now, any impl of the data layout importer will need to load a backend, so that's a hard dependency on the mechanism. I would argue we should split some target information from the LLVM backend, but that's another conversation.

rolfmorel replied:
As @fabianmcg says, I think we need the promised interface to make it optional to load/link the `mlir/lib/target` lib with its dependency on LLVM libs. If the `-llvm-data-layout-from-target` (or `-llvm-target-to-data-layout`) pass would access LLVM's TargetMachine and DataLayout directly, where should it live? Not in `Dialect/LLVMIR/Transforms`, right? Can passes live inside `mlir/lib/Target`? Is there another natural way of silo-ing off the pass so people optionally incur the LLVM dependency?

> Do we anticipate other LLVMTargetAttr to provide this interface?

In my mind `#nvvm.target` and `#rocdl.target` should be implementers and would essentially be specialized versions of `#llvm.target` (as for these backends we do happen to have dialects).

banach-space commented:
> Arguably we don't usually care about armv7 or lower, but I'll let @banach-space comment on that.

Yes, it's fine if we focus on ArmV8 and higher.

rolfmorel commented on mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:
```diff
@@ -1304,6 +1305,128 @@ def LLVM_TargetFeaturesAttr : LLVM_Attr<"TargetFeatures", "target_features">
+  let description = [{
+    An attribute to hold LLVM target information, specifying LLVM's target
+    `triple` string, the target `chip` string (i.e. the `cpu` string), and
+    target `features` string as an attribute. The latter two are optional.
```
Up above it is argued that `chip` (i.e. the `cpu` string) shouldn't be optional. I don't mind either way.

gysit replied:
Yeah that works as well. I was just commenting because the chip is not marked optional in the parameters list on line 1331 below.

rengolin replied:
Yes, seems we all agree the `chip` part needs to be there, non-empty. Without this, it's really hard to build a reasonable target machine, and LLVM usually has "perfect" knowledge of the target features for its supported targets.

Target features are useful for creating non-standard LLVM configurations OR when LLVM doesn't have "perfect" knowledge of the environment (which can be controlled by external flags like environment variables or wrapper scripts).

rolfmorel commented on mlir/test/Dialect/LLVMIR/data-layout-from-target-invalid.mlir:
```diff
@@ -0,0 +1,9 @@
+// REQUIRES: target=x86{{.*}}
+// RUN: mlir-opt %s -llvm-data-layout-from-target --split-input-file --verify-diagnostics
+
+// expected-error @+1 {{failed to obtain llvm::DataLayout from #llvm.target}}
+module attributes { dlti.dl_spec = #dlti.dl_spec<index = 32>,
+llvm.target =
+    #llvm.target<triple="x64_86-unknown-linux",
+                 chip="lakesky"> } {
+}
```
Yep. Will change it to something more obvious. (How about "CPU_THAT_DOES_NOT_EXIST"?)

rolfmorel commented on mlir/include/mlir/Dialect/LLVMIR/DataLayoutImporter.h:
```diff
@@ -38,23 +38,30 @@ namespace detail {
-/// entries from the default data layout specification found in the language
-/// reference (https://llvm.org/docs/LangRef.html#data-layout) if they are not
-/// overwritten by the provided data layout.
+/// Helper class that translates an LLVM data layout string to an MLIR data
+/// layout specification. Only integer, float, pointer, alloca memory space,
```
There's this page: https://mlir.llvm.org/docs/DataLayout/#data-layout-specifications

It doesn't do a good job of specifying all the properties that are supported. Nor does it clearly explain how those properties are encoded in `dl_spec`s.

Is there any consensus on whether tracking issues on Github are worth the effort?

rolfmorel commented on mlir/include/mlir/Dialect/LLVMIR/DataLayoutImporter.h:
```diff
@@ -38,23 +38,30 @@ namespace detail {
-    translateDataLayout(llvmDataLayout);
+  DataLayoutImporter(MLIRContext *context, StringRef dataLayoutStr)
+      : dataLayoutStr(dataLayoutStr), context(context) {
+    // Append the default data layout string specified in the language reference
```
It is indeed opaque that this is appending is/was happening in `dataLayoutSpecFromDataLayoutStr()` (formerly `translateDataLayout()`). Will address this in my next revision.

rengolin commented on mlir/lib/Dialect/LLVMIR/Transforms/DataLayoutFromTarget.cpp:
```diff
@@ -0,0 +1,61 @@
+      auto dataLayoutAttr = LLVM::DataLayoutAttr::get(
+          &getContext(), dataLayout->getStringRepresentation());
+
+      StringRef dlSpecIdentifier = "dlti.dl_spec";
```
Yes, similar to @joker-eph comment above, no magic strings.

krzysz00 reviewed (commented):
Triviality I forgot to actually submit as a review comment weeks ago, actual review .. maybe next week, though it looks like it's been pretty well covered

Commit:
Address remaining comments

Commit:
Style improvements

Commit:
Further minor fixes

joker-eph commented on mlir/lib/Target/LLVMIR/Transforms/TargetToDataLayout.cpp:
```diff
@@ -0,0 +1,104 @@
+  const llvm::Target *target =
+      llvm::TargetRegistry::lookupTarget(triple, error);
+  if (!target || !error.empty()) {
+    LLVM_DEBUG({
+      llvm::dbgs() << "Looking up target '" << triple << "' failed: " << error
+                   << "\n";
+    });
```
```suggestion
    LDBG() <<  "Looking up target '" << triple << "' failed: " << error;
```

joker-eph reviewed (approved):
LGTM overall, but let's wait on @fabianmcg review as well.

krzysz00 reviewed (commented):
Is hooking up this to the LLVM lowerings - so thaw we can, for example, know what the actual size of an `[N x i16]` is - a future PR?

fabianmcg reviewed (approved):
Thank you, LGTM! See the minor comments. Can you  build the patch with `BUILD_SHARED_LIBS=ON` and confirm there are no linking issues? That way we can catch them before merging.

Commit:
Address Mehdi's and Fabian's comments

rolfmorel commented on mlir/lib/Target/LLVMIR/Transforms/TargetToDataLayout.cpp:
```diff
@@ -0,0 +1,104 @@
+    auto targetAttr = op->getAttrOfType<LLVM::TargetAttrInterface>(
+        LLVM::LLVMDialect::getTargetAttrName());
+    if (!targetAttr)
+      return;
```
Now that the pass operates on the target op directly, I think it should be a failure indeed. Changed.

Commit:
Merge remote-tracking branch 'origin/main' into llvm-target-dlti

Commit:
Final fixes

Commit:
Fix formatting

Commit:
Merge remote-tracking branch 'origin/main' into llvm-target-dlti

Commit:
Add back in LLVMTypes.h as include in LLVMAttrs.h

As of course there are people that depend on it...
AMX Dialect and Flang ... and potentially downstreams

Commit:
Reinstate mlir::translateDataLayout

As it is used, e.g., by Flang

Commit:
Fix formatting
